backends/ent/store.go

```diff
@@ -0,0 +1,479 @@
+		}
+	}
+
+	if _, ok := opts.BackendOptions.(*BackendOptions); !ok {
+		return fmt.Errorf("%w", errInvalidEntOptions)
+	}
+
+	backend.init(backend.BackendOptions)
+
+	defer backend.client.Close()
+
+	if err := backend.StoreMetadata(doc.Metadata); err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
+	if err := backend.StoreNodeList(doc.NodeList); err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
+	entNodeList, err := backend.client.NodeList.Query().
+		Where(
+			nodelist.Or(
+				nodelist.HasDocumentWith(document.HasMetadataWith(metadata.IDEQ(doc.Metadata.Id))),
+				nodelist.Not(nodelist.HasDocument()),
+			)).
+		Only(backend.ctx)
+	if err != nil {
+		return fmt.Errorf("%w", err)
+	}
+
+	err = backend.client.Document.Create().
+		SetMetadataID(doc.Metadata.Id).
+		SetNodeListID(entNodeList.ID).
+		OnConflict().
+		Ignore().
+		Exec(backend.ctx)
+	if err != nil && !ent.IsConstraintError(err) {
+		return fmt.Errorf("ent.Document: %w", err)
+	}
+
+	return nil
+}
+
+func (backend *Backend) StoreDocumentTypes(docTypes []*sbom.DocumentType) error {
+	// Set up client if this method was called directly.
+	if backend.BackendOptions == nil {
+		backend.BackendOptions = NewBackendOptions()
+		defer backend.client.Close()
+	}
+
+	backend.init(backend.BackendOptions)
+
+	for _, dt := range docTypes {
+		typeName := documenttype.Type(dt.String())
+
+		newDocType := backend.client.DocumentType.Create().
+			SetNillableType(&typeName).
+			SetNillableName(dt.Name).
+			SetNillableDescription(dt.Description)
+
+		if backend.metadataID != "" {
+			newDocType.SetMetadataID(backend.metadataID)
+		}
+
+		err := newDocType.OnConflict().Ignore().Exec(backend.ctx)
+		if err != nil && !ent.IsConstraintError(err) {
+			return fmt.Errorf("ent.DocumentType: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (backend *Backend) StoreExternalReferences(refs []*sbom.ExternalReference) error {
+	// Set up client if this method was called directly.
+	if backend.BackendOptions == nil {
+		backend.BackendOptions = NewBackendOptions()
+		defer backend.client.Close()
+	}
+
+	backend.init(backend.BackendOptions)
+
+	for _, ref := range refs {
+		newRef := backend.client.ExternalReference.Create().
+			SetURL(ref.Url).
+			SetComment(ref.Comment).
+			SetAuthority(ref.Authority).
+			SetType(externalreference.Type(ref.Type.String()))
+
+		if backend.nodeID != "" {
+			newRef.SetNodeID(backend.nodeID)
+		}
+
+		id, err := newRef.OnConflict().Ignore().ID(backend.ctx)
+		if err != nil && !ent.IsConstraintError(err) {
+			return fmt.Errorf("ent.ExternalReference: %w", err)
+		}
+
+		backend.externalReferenceID = id
+
+		if err := backend.StoreHashesEntries(ref.Hashes); err != nil {
+			return fmt.Errorf("%w", err)
+		}
+	}
+
+	return nil
+}
+
+func (backend *Backend) StoreHashesEntries(hashes map[int32]string) error {
+	// Set up client if this method was called directly.
+	if backend.BackendOptions == nil {
+		backend.BackendOptions = NewBackendOptions()
+		defer backend.client.Close()
+	}
```

Comment:
Maybe I'm missing some pattern but IMHO, since this is a method of `Backend`, all of these initializations should happen when the backend is created. If ww reach this point and the backend state is not consistent the methods using the client should err, especially to avoid an unexpected behavior by running the method with a new set of options created inside the method. In other words, the `backend.init()` on the next line should also not happen here.

Reply:
Will rework this

Reply:
See https://github.com/protobom/storage/pull/4#issuecomment-2147895186 for how this was reworked